src/huggingface_hub/repository.py

```diff
@@ -148,73 +150,123 @@ def clone_from(self, repo_url: str, use_auth_token: Union[bool, str, None] = Non
                     cwd=self.local_dir,
                 )
             else:
-                logger.warning(
-                    "[Repository] local_dir is not empty, so let's try to pull the remote over a non-empty folder."
-                )
-                subprocess.run(
-                    "git init".split(),
-                    stderr=subprocess.PIPE,
-                    stdout=subprocess.PIPE,
-                    check=True,
-                    encoding="utf-8",
-                    cwd=self.local_dir,
+                # Check if the folder is the root of a git repository
+                in_repository = (
+                    os.path.exists(os.path.join(self.local_dir, ".git"))
+                    and subprocess.run(
+                        "git branch".split(), cwd=self.local_dir
+                    ).returncode
+                    == 0
                 )
 
-                output = subprocess.run(
-                    "git remote -v".split(),
-                    stderr=subprocess.PIPE,
-                    stdout=subprocess.PIPE,
-                    check=True,
-                    encoding="utf-8",
-                    cwd=self.local_dir,
-                )
-
-                if "origin" not in output.stdout.split():
+                if in_repository:
+                    remotes = subprocess.run(
+                        "git remote -v".split(),
+                        stderr=subprocess.PIPE,
+                        stdout=subprocess.PIPE,
+                        check=True,
+                        encoding="utf-8",
+                        cwd=self.local_dir,
+                    ).stdout
+
+                    # Remove token for the test with remotes.
+                    remote_url = re.sub(r"https://.*@", "https://", repo_url)
+                    if remote_url in remotes.split():
```

Comment:
Porting my comment from https://github.com/huggingface/transformers/pull/12187:

Btw here I'm checking the remotes to see if a repository is linked to another, but we could theoretically imagine two repositories with dissimilar remotes that would still be linked by history.

A better way to check this would be to check the commit history directly and check if they align - but I found the remote check simpler, and I believe it's enough of a check for our use-case.

Reply:
yes definitely good enough